examples/nlp/summarization.py

```diff
@@ -0,0 +1,378 @@
+dataset with the `collator` defined above. The method expects certain parameters:
+
+- **columns**: the columns which will serve as our independant variables
+- **batch_size**: our batch size for training
+- **shuffle**: whether we want to shuffle our dataset
+- **collate_fn**: our collator function
+
+Additionally, we also define a relatively smaller `generation_dataset` to calculate
+`ROUGE` scores on the fly while training.
+"""
+
+train_dataset = tokenized_datasets["train"].to_tf_dataset(
+    batch_size=BATCH_SIZE,
+    columns=["input_ids", "attention_mask", "labels"],
+    shuffle=True,
+    collate_fn=data_collator,
+)
+test_dataset = tokenized_datasets["test"].to_tf_dataset(
+    batch_size=BATCH_SIZE,
+    columns=["input_ids", "attention_mask", "labels"],
+    shuffle=False,
+    collate_fn=data_collator,
+)
+generation_dataset = (
+    tokenized_datasets["test"]
+    .shuffle()
+    .select(list(range(200)))
+    .to_tf_dataset(
+        batch_size=BATCH_SIZE,
+        columns=["input_ids", "attention_mask", "labels"],
+        shuffle=False,
+        collate_fn=data_collator,
+    )
+)
+
+"""
+## Building and Compiling the the model
+"""
+
+"""
+Now we will define our optimizer and compile the model. The loss calculation is handled
+internally and so we need not worry about that!
+"""
+
+from transformers import AdamWeightDecay
+
+optimizer = AdamWeightDecay(learning_rate=LEARNING_RATE, weight_decay_rate=WEIGHT_DECAY)
+model.compile(optimizer=optimizer)
+
+"""
+## Training and Evaluating the model
+"""
+
+"""
+To evaluate our model on-the-fly while training, we will defien `metric_fn` which will
+calculate the `ROUGE` score between the groud-truth and predictions.
+"""
+
+
+def metric_fn(eval_predictions):
+    predictions, labels = eval_predictions
+    decoded_predictions = tokenizer.batch_decode(predictions, skip_special_tokens=True)
+    for label in labels:
+        label[label < 0] = tokenizer.pad_token_id  # Replace masked label tokens
+    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
+    # Rouge expects a newline after each sentence
+    decoded_predictions = [
+        "\n".join(nltk.sent_tokenize(pred.strip())) for pred in decoded_predictions
+    ]
+    decoded_labels = [
+        "\n".join(nltk.sent_tokenize(label.strip())) for label in decoded_labels
+    ]
+    result = metric.compute(
+        predictions=decoded_predictions, references=decoded_labels, use_stemmer=True
+    )
+    # Extract a few results
+    result = {key: value.mid.fmeasure * 100 for key, value in result.items()}
+    # Add mean generated length
+    prediction_lens = [
+        np.count_nonzero(pred != tokenizer.pad_token_id) for pred in predictions
+    ]
+    result["gen_len"] = np.mean(prediction_lens)
+
+    return result
+
+
+"""
+Now we can finally start training our model!
+"""
+
+from transformers.keras_callbacks import KerasMetricCallback
+
+metric_callback = KerasMetricCallback(
```

Comment:
Q: instead of using a callback for Rouge, can you use the KerasNLP metric? You'd just pass it to `compile()`. https://keras.io/api/keras_nlp/metrics/

Reply:
Did you figure this out?

Reply:
Hi @fchollet ,

First of all, thank you so much for your detailed feedback and review on all my PRs. This looks a bit tricky as passing Rouge from KerasNLP metric to `compile()` would require my model inputs and targets to be strings which is not the case in my current implementation (my current implementation works on input ids). So if I want to use Rouge from KerasNLP metric, I have two options:

1) Simply replace [this line](https://github.com/keras-team/keras-io/pull/967/files#diff-2deb36c911c9b27529bce55cb6f0729bbb9d0f27a0ba0c0f137659538ae3fbe2R310-R312) with rouge from KerasNLP. This currently uses rouge from Hugging Face. This will keep the callback as it is.

2) Write a custom evaluation metric that decodes my model to strings similar and use KerasNLP Rouge in that metric, something similar to my WER implementation [here](https://github.com/keras-team/keras-io/pull/996/files#diff-e8a623de2815fff9118c26111ecf0fa511df13fb0eb8b19222f1172aa2aac28dR594-R603).